sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -1880,6 +1885,17 @@ class DatasetSuite extends QueryTest with SharedSparkSession {
     }
     assert(exchanges.size == 2)
   }
+
+  test("tail with different numbers") {
+    Seq(0, 2, 5, 10, 50, 100, 1000).foreach { n =>
+      assert(spark.range(n).tail(5) === (math.max(n - 5, 0) until n))
+    }
+  }
+
+  test("tail should not accept minus value") {
```

Comment:
do we have a nice error message for `df.tail(...).otherOperation.collect`?

Reply:
`tail` itself already collects so other operations can't be chained.